Support EFS attachment in lambda function configuration

pgrzesik opened the pull request:
~~As of right now the PR is in a WIP/draft stage - I wanted to discuss the implementation decision as a part of this issue and I figured it will be easier to be able to reference code with the selected approach.~~

UPDATE:
Hello, thanks a lot for previous comments, I prepared the PR for review, however, there are still a few things that I'm not sure about:

1. What is the proper way of testing the schemas? I wanted to test scenarios where invalid configs are provided but I've kept running into schema validation errors which aren't easy to catch and assert against.
2. What is your recommended way of defining ARNs patterns in schema? I've seen some regex examples in `arnRegularExpressions.js` but it seems like they're a bit more "basic" versions than what is expected by CloudFormation for some parameters. Do you recommend to move the patterns for ARNs to a separate file/module for potential reuse?
3. Should we validate that the function has VPC configured when someone configures EFS? It is a requirement and the deployment will fail, but I'm wondering what's the approach to such situations in general.
4. Should we include the note about requirement for `vpc` in the documentation for `efs`?

In addition - I keep running into errors on CI which are not present on my local env - do you have an idea of what might be causing it? I have a suspicion that it might be related to parallelization/test isolation as I'm reusing the same fixture multiple times, but I'm not sure what would be the recommended way to remedy that issue.

Thanks a ton in advance for taking your time to take a look at this PR :bow:

<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on solution in corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/master/tests/README.md
-->

<!-- ⚠️⚠️ Ensure that support for Node.js v6 is maintained. -->

<!--
⚠️⚠️ Ensure that proposed change passes CI. Confirm on that by running following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/master/tests/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide link to corresponding issue

• If PR *partially* addresses issue, ensure to rename "Closes" to "Addresses" ("Closes" term will automatically close an issue on PR merge)
• If it's a documentation update or obvious bug fix that has no corresponding issue, replace this line with short description of made changes
-->

Closes: #7848

medikoo commented on lib/plugins/aws/package/compile/functions/index.js:
```diff
@@ -301,6 +301,71 @@ class AwsCompileFunctions {
+
+      // TODO: Simplify these conditions
+      if (fileSystemConfigs && fileSystemConfigs.length !== 0) {
+        if (fileSystemConfigs.length > 1) {
+          // TODO: Add explanation - CloudFormation limitation
+          const errorMessage = 'You can specify at most 1 fileSystemConfig';
+          throw new this.serverless.classes.Error(errorMessage);
```
Let's not put any validation here.

We've just added schema based config validation. So we should simply configure a JSON schema for those properties here: https://github.com/serverless/serverless/blob/d403d9c6d3dccdfc9f79d7c42b3cbc1a06fcfac0/lib/plugins/aws/provider/awsProvider.js#L145

medikoo commented on lib/plugins/aws/package/compile/functions/index.js:
```diff
@@ -301,6 +301,71 @@ class AwsCompileFunctions {
+          // TODO: extract it as a util as it's used in multiple places?
+          // update the PolicyDocument statements (if default policy is used)
+          if (iamRoleLambdaExecution) {
+            iamRoleLambdaExecution.Properties.Policies[0].PolicyDocument.Statement = _.unionWith(
```
I think `unionWith` is unjustified here (there's no chance in some other part we're adding exactly same statement. Let's just unconditonally push it

pgrzesik commented:
Hello, thanks a lot for previous comments, I prepared the PR for review, however, there are still a few things that I'm not sure about:

1. What is the proper way of testing the schemas? I wanted to test scenarios where invalid configs are provided but I've kept running into schema validation errors which aren't easy to catch and assert against.
2. What is your recommended way of defining ARNs patterns in schema? I've seen some regex examples in `arnRegularExpressions.js` but it seems like they're a bit more "basic" versions than what is expected by CloudFormation for some parameters. Do you recommend to move the patterns for ARNs to a separate file/module for potential reuse?
3. Should we validate that the function has VPC configured when someone configures EFS? It is a requirement and the deployment will fail, but I'm wondering what's the approach to such situations in general.
4. Should we include the note about requirement for `vpc` in the documentation for `efs`?

In addition - I keep running into errors on CI which are not present on my local env - do you have an idea of what might be causing it? I have a suspicion that it might be related to parallelization/test isolation as I'm reusing the same fixture multiple times, but I'm not sure what would be the recommended way to remedy that issue.

Thanks a ton in advance for taking your time to take a look at this PR :bow:

medikoo commented:
Hey @pgrzesik, it's all great questions. See my answers:

> What is the proper way of testing the schemas?

Let's do not test schemas (schemas engine is extensively tested on its own and we do not need to double confirm that it works, also we already have a sanity checks that confirm that validation works in general and that's good enough).

Let's just test valid scenarios.

> What is your recommended way of defining ARNs patterns in schema? 

Refer to https://github.com/serverless/serverless/blob/4660acd324cfa6c786245ea581691a23758ce960/lib/configSchema.js#L116-L119 (it's e.g. used here: https://github.com/serverless/enterprise-plugin/blob/25094ebd02e6ae350d2055f9dfbde241cefd1762/lib/plugin.js#L84 )

If you feel, we need to a more precise (or other) definitions feel free to add them to `configSchema.js`

> Should we validate that the function has VPC configured when someone configures EFS?

It's a good question. I think we should (assuming we agree that it's only through Framework config  deployed functions have vps configured)

> Should we include the note about requirement for vpc in the documentation for efs

We may, but in first place it'll be good to link general AWS doc on EFS.

(_I'm on holiday this week, so I may not respond promptly during that time_)

pgrzesik commented:
Thanks @medikoo for your responses :bow: 

As for the VPC validation - I decided to go the route with validating that as a part of `functions/index.js` as defining schema that can support dependencies on either `provider` or `functions` properties seems outside of the scope of this PR - unless there's a way to define that in a reasonable way - I would love to learn if there's an option to do that. Or if in such a situation you'd prefer to keep the VPC validation out of the scope.

During testing this feature, I've stumbled upon a few errors with `Uncaught ServerlessError: Invalid variable reference syntax for variable AWS::Partition.` - they are visible in failed CI tests - do you know by any chance what might be causing them?

Last thing: What is your stance on writing integration tests for this one? I believe it would be great to have integration tests for this one, however, it requires to create a lot of resources (VPC, EFS) via CloudFormation to test it and I didn't see any other integration tests heavily depending on pure CloudFormation resource creation - what is your opinion here? Go for it or skip integration tests for this one?

Hopefully, you have great vacations and I don't hear anything from you till next week :raised_hands:

pgrzesik commented:
Hello :wave: I've implemented the VPC validation as mentioned in my previous message. The two below questions still stand:
1. What is your stance on writing integration tests for this one? I believe it would be great to have integration tests for this one, however, it requires to create a lot of resources (VPC, EFS) via CloudFormation to test it and I didn't see any other integration tests heavily depending on pure CloudFormation resource creation - what is your opinion here? Go for it or skip integration tests for this one?
2. During testing this feature, I've stumbled upon a few errors with Uncaught ServerlessError: Invalid variable reference syntax for variable AWS::Partition. - they are visible in failed CI tests - do you know by any chance what might be causing them?

Thanks for your time,
Piotr

medikoo commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -142,7 +142,39 @@ class AwsProvider {
+                  localMountPath: { type: 'string', pattern: '^/mnt/[a-zA-Z0-9-_.]+$' },
+                  arn: {
+                    oneOf: [
+                      {
+                        $ref: '#/definitions/awsArn',
+                      },
```
AWS provides quite specific pattern for this ARN: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-filesystemconfig.html

Let's maybe the configure it in place with it. It'll ensure better validation

pgrzesik replied:
Updated :+1:

medikoo commented on lib/plugins/aws/package/compile/functions/index.js:
```diff
@@ -397,6 +397,48 @@ class AwsCompileFunctions {
 
+      const fileSystemConfigs = functionObject.fileSystemConfigs;
+
+      if (Array.isArray(fileSystemConfigs) && fileSystemConfigs.length === 1) {
```
At this point, as we're backed by schema validation, I would just assume that if `fileSystemConfigs` it is an array, and has at most length 1.

Therefore I'd write it as `if (fileSystemConfigs.length)`

pgrzesik replied:
I still kept the `Array.isArray` check here, because if I understand it correctly, schema is not enforced at the moment and for example in tests, with missing `fileSystemConfigs` it's erroring out as it tries to check length of `undefined`. Please let me know what are your thoughts on this

medikoo replied:
> schema is not enforced at the moment

That's true, but we also live with an agreement, that if schema is not satisfied then things can break badly. Code is not bulletproof to accept any kind of input in config, and may crash with non informative programmer errors (like `x.x.x is undefined`) - and it was always like that.

Also with one of the major upgrades we will crash by default when schema is not satisfied (warn by default is transitory state)

Therefore to simplify maintenance let's blindly assume that at this point configuration adheres to schema. I believe it's totally fine, and let us deal with less footprint

medikoo commented on lib/plugins/aws/package/compile/functions/index.js:
```diff
@@ -397,6 +397,48 @@ class AwsCompileFunctions {
+          );
+        }
+
+        const shouldUpdateIamRole =
+          typeof fileSystemConfig.arn === 'string' && fileSystemConfig.arn.startsWith('arn');
+        if (shouldUpdateIamRole) {
+          const iamRoleLambdaExecution = cfTemplate.Resources.IamRoleLambdaExecution;
+
+          const stmt = {
+            Effect: 'Allow',
+            Action: ['elasticfilesystem:ClientMount', 'elasticfilesystem:ClientWrite'],
+            Resource: [fileSystemConfig.arn],
+          };
+
```
I think we should require arn string unconditionally, as with CloudFormation instructions we're unable to setup proper access rights.

We already do that in many other cases, so i think such restriction is not that bad (also in case of EFS it shouldn't be that problematic to rely only on ARN strings I think)

pgrzesik replied:
So, do you suggest to totally skip support for `Fn::Import`, `Ref` and `Fn::GetAtt`?

medikoo replied:
Yes, however if you see some significant issue in doing that, please elaborate

pgrzesik replied:
That sounds reasonable, it's a bit clunkly to support it because, as you mentioned, it forces you to use custom Lambda Role with EFS policies, otherwise the function won't be able to interact with EFS at all. I removed support for it in the current iteration

BrentFarese replied:
@pgrzesik here in 2024, what is the recommended way to use `Ref` or `Fn::GetAtt` here in `fileSystemConfig.arn`?  We are creating EFS as part of our `serverless.ts` (see below), but cannot reference the created `arn`.  Seems very limiting now in 2024:

```
const serverlessConfiguration = {
  custom: {
    esbuild: {
      external: ['canvas'],
      minify: true,
      packager: 'yarn',
      treeShaking: true,
    },
    localstack: {
      debug: false,
      edgePort: 4566,
      host: 'http://localhost',
      lambda: {
        mountCode: false,
      },
      stages: ['local'],
    },
    sentry: {
      dsn: serverlessEnv.SENTRY_DSN_LAMBDA,
    },
  },
  frameworkVersion: '3',
  functions: {
    clearDocumentSignatures: {
      handler: 'src/functions/clearDocumentSignatures.clearDocumentSignatures',
      name: 'clearDocumentSignatures-${opt:stage}',
      timeout: 600,
    },
    createDocumentVersions: {
      events: [
        { sqs: { arn: { 'Fn::GetAtt': ['CreateDocumentVersionQueue', 'Arn'] }, batchSize: 10 } },
      ],
      handler: 'src/functions/createDocumentVersions.createDocumentVersions',
      name: 'createDocumentVersions-${opt:stage}',
      timeout: 600,
    },
    extractFileJSON: {
      dependsOn: ['FileSystem'],
      fileSystemConfig: {
        arn: 'FileSystem.arn',
        localMountPath: '/mnt/tmp',
      },
      handler: 'src/functions/extractFileJSON.extractFileJSON',
      name: 'extractFileJSON-${opt:stage}',
      timeout: 600,
    },
    flushDocumentVersions: {
      handler: 'src/functions/flushDocumentVersions.flushDocumentVersions',
      name: 'flushDocumentVersions-${opt:stage}',
      timeout: 600,
    },
    scheduleCreateDocumentHistoryVersions: {
      environment: {
        // Seems like this injects CREATE_DOCUMENT_VERSION_QUEUE_URL onto process.env...
        CREATE_DOCUMENT_VERSION_QUEUE_URL: { Ref: 'CreateDocumentVersionQueue' },
      },
      events: [{ schedule: { rate: 'rate(1 minute)' } }],
      handler:
        'src/functions/scheduleCreateDocumentHistoryVersions.scheduleCreateDocumentHistoryVersions',
      name: 'scheduleCreateDocumentHistoryVersions-${opt:stage}',
      timeout: 600,
    },
    setDocumentState: {
      handler: 'src/functions/setDocumentState.setDocumentState',
      name: 'setDocumentState-${opt:stage}',
      timeout: 600,
    },
  },
  plugins: ['serverless-esbuild', 'serverless-localstack', 'serverless-sentry'],
  provider: {
    environment: serverlessEnv,
    name: 'aws',
    runtime: 'nodejs16.x',
    vpc: {
      securityGroupIds: isNoVPCNeeded ? [] : securityGroupIds,
      subnetIds: isNoVPCNeeded ? [] : subnetIds,
    },
  },
  resources: {
    Resources: {
      CreateDocumentVersionQueue: {
        Properties: { QueueName: 'create-document-version-${opt:stage}', VisibilityTimeout: 3600 },
        Type: 'AWS::SQS::Queue',
      },
      FileSystem: { Properties: {}, Type: 'AWS::EFS::FileSystem' },
      FileSystemAccessPoints: {
        DependsOn: 'FileSystemMountTarget',
        Properties: {
          FileSystemId: { Ref: 'FileSystem' },
          Path: '/',
          PosixUser: { Gid: '1000', Permissions: 777, Uid: '1000' },
          RootDirectory: { CreationInfo: { OwnerGid: '1000', OwnerUid: '1000', Permissions: 755 } },
        },
        Type: 'AWS::EFS::AccessPoint',
      },
      FileSystemMountTarget: {
        Properties: {
          FileSystemId: { Ref: 'FileSystem' },
          SecurityGroups: securityGroupIds,
          SubnetId: serverlessEnv.AWS_SUBNET_PUBLIC_ID,
        },
        Type: 'AWS::EFS::MountTarget',
      },
    },
  },
  service: 'collab-worker',
} satisfies Serverless;
```

medikoo commented on lib/plugins/aws/package/compile/functions/index.test.js:
```diff
@@ -2842,4 +2842,161 @@ describe('AwsCompileFunctions #2', () => {
+      const arn =
+        'arn:aws:elasticfilesystem:region:111111111111:access-point/fsap-a1a1a1a1a1a1a1a1a';
+      const localMountPath = '/mnt/path';
+      runServerless({
```
For good tests efficiency, it'll be nice to test all _success_ scenarios with one `runServerless` run, as e.g. we do in case of ALB health check here: https://github.com/serverless/serverless/blob/04d463f4e438d02fef958fce34677a06645fee82/lib/plugins/aws/package/compile/events/alb/lib/healthCheck.test.js#L9-L113

Also same way I'd rather reuse the `function` fixture. At least when we can easily create a case purely by extension (and we do not need to reuse it in other tests files) it's good to keep fixtures count minimal

pgrzesik replied:
makes perfect sense, I will adjust it :+1:

medikoo commented on lib/plugins/aws/package/compile/functions/index.test.js:
```diff
@@ -2842,4 +2842,161 @@ describe('AwsCompileFunctions #2', () => {
+      const arn =
+        'arn:aws:elasticfilesystem:region:111111111111:access-point/fsap-a1a1a1a1a1a1a1a1a';
+      const localMountPath = '/mnt/path';
+      runServerless({
```
We should return the result of `runServerless` otherwise orphaned async flow is created, and it might be responsible for crash in tests

pgrzesik replied:
thanks - I totally missed that and it seems like it might be the cause of errors :crossed_fingers:

pgrzesik replied:
That was exactly the case, thanks again :raised_hands:

tommedema commented:
Will this take care of creating an EFS if not specified? And does this require the lambda to be in a VPC?

pgrzesik commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -142,7 +142,39 @@ class AwsProvider {
+                      {
+                        $ref: '#/definitions/awsArn',
+                      },
+                      {
+                        type: 'object',
+                        propertyNames: {
+                          pattern: '^(Fn::ImportValue|Fn::GetAtt|Ref)$',
+                        },
+                        minProperties: 1,
+                        maxProperties: 1,
+                      },
```
That's a great idea - the question here is if we want to provide support for `Fn::ImportValue`, `Fn::GetAtt` and `Ref` at all (see my other response)

pgrzesik replied:
I removed this configuration altogether as it was no longer needed with keeping support for only ARNs so I felt it doesn't really belong to this PR anymore

pgrzesik commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -142,7 +142,39 @@ class AwsProvider {
+                        propertyNames: {
+                          pattern: '^(Fn::ImportValue|Fn::GetAtt|Ref)$',
+                        },
+                        minProperties: 1,
+                        maxProperties: 1,
```
That's much better :+1:

pgrzesik commented on docs/providers/aws/guide/functions.md:
```diff
@@ -501,3 +501,25 @@ functions:
+
+## EFS Configuration
+
+You can use [Amazon EFS with Lambda](https://docs.aws.amazon.com/lambda/latest/dg/services-efs.html) by adding a `fileSystemConfigs` property in the function configuration in `serverless.yml`. `fileSystemConfigs` should be a list of objects that contain the `arn` and `localMountPath` properties. The `arn` property should reference an existing EFS Access Point, where the `localMountPath` should sp ...
```
At the moment it's possible to attach only one fileSystem to Lambda and I'm not sure when/if that's going to change in the future. In CloudFormation, they're specified as an array with maximum of 1 (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-filesystemconfigs), most likely as a way to be future-proof when (potentially) it would be possible to attach more than one filesystem to Lambda. It might make sense in case of `serverless` to have this as singular `fileSystemConfig`, however, what do you think the migration in the future should look like if it would become possible to configure more than one filesystem?

medikoo replied:
Let me check with AWS what's the probability of having that lifted soon (I'll get back once I have an answer.. probably in 24-48 hours timeframe)

pgrzesik replied:
Thank you :bow:

medikoo replied:
@pgrzesik I didn't receive any specific answer.

I'd say let's make it singular, so `fileSystemConfig` (not `fileSystemConfigs`), and if eventually AWS will shift the limit, we will introduce `fileSystemConfigs`, at this point I doubt it'll be the case anytime soon.

pgrzesik commented:
@medikoo Thanks a ton for all your time and responses. I will implement the suggested changes + attempt to introduce an integration test for this functionality.

medikoo commented:
> Will this take care of creating an EFS if not specified? 

@tommedema No, in this proposal we assume it was setup externally. Do you feel we should support creation of file systems together with lambda?

> And does this require the lambda to be in a VPC?

AFAIK it's a requirement on AWS side

tommedema commented:
> > Will this take care of creating an EFS if not specified?
> 
> @tommedema No, in this proposal we assume it was setup externally. Do you feel we should support creation of file systems together with lambda?

Would be super useful, though complicated because you'd have to create a VPC and deal with ENIs etc

medikoo commented:
> Would be super useful, though complicated because you'd have to create a VPC and deal with ENIs etc

@tommedema we've once attempted to bring VPC support (see https://github.com/serverless/serverless/pull/6675) still we aborted work on that, as we agreed it's not a thing that should be setup together with a service (usually VPC is setup with intention to be used by many services), therefore it doesn't really belong to `serverless.yml`. 

I take same assumption can be made about EFS (?)

pgrzesik commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -142,7 +142,39 @@ class AwsProvider {
+          properties: {
+            handler: { type: 'string' },
+            fileSystemConfigs: {
+              description:
```
Updated :+1:

pgrzesik commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -142,7 +142,39 @@ class AwsProvider {
+              description:
+                'Parameters describing connection between a Lambda function and an Amazon EFS file system',
+              type: 'array',
+              minItems: 1,
```
Updated :+1:

pgrzesik commented:
@medikoo I believe I applied all the suggested changes and I'm waiting on a response in regard to potentially supporting multiple `fileSystemConfigs` in the future. I'll also try to write a proper integration test and will report back with how efficient that is. 

As for the VPC @tommedema - it is a requirement on AWS side, it's impossible to setup EFS with Lambda if Lambda is not in a VPC. 

As for creating EFS resources as a part of serverless - I think it might be a valid approach, the only thing I'm worried about is how much customization should be allowed in such situations and how that should be exposed to users.

medikoo commented:
> @medikoo I believe I applied all the suggested changes and I'm waiting on a response in regard to potentially supporting multiple fileSystemConfigs in the future. 

Thank you @pgrzesik ! I contacted AWS, but at this point answer I got, was "I don't know but I'll ask engineering team and let you know". I propose to wait  until end of this week, and then deal with it even if we don't have an answer at this point.

pgrzesik commented:
Hello @medikoo :wave: As suggested, I changed the implementation to use singular `fileSystemConfig` at the moment.

As mentioned previously, I also added an integration test. I experimented a bit and decided that the CF approach is the most reasonable one (at least in my opinion). I noticed that in some cases, the first invoke in test was failing, but all subsequent invokes were successful, hence the "retry" functionality - I tried to follow a similar convention as in `integration-all/api-gateway/tests.js`. I ran the test ~20 times and the average time it takes to run the test (including provisioning & deprovisioning) is about 10 minutes (maximum I got was 11 minutes, lowest 9 minutes). Please let me know if that's acceptable for it to become a part of standard integration tests suite.

Thank you for your patience :bow:

medikoo commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -220,7 +220,22 @@ class AwsProvider {
+                arn: {
+                  type: 'string',
+                  pattern:
+                    '^arn:aws[a-zA-Z-]*:elasticfilesystem:[a-z]{2}((-gov)|(-iso(b?)))?-[a-z]+-[1-9]{1}:[0-9]{12}:access-point/fsap-[a-f0-9]{17}',
```
I think we need to add `$` at the end (to not allow any extra trailing chars)

medikoo commented on tests/integration-all/fileSystemConfig/service/serverless.yml:
```diff
@@ -0,0 +1,12 @@
+service: CHANGE_TO_UNIQUE_PER_RUN
+
```
Let's also ensure `configValidationMode: error` (as in other service fixtures)

medikoo commented on tests/integration-all/fileSystemConfig/service/serverless.yml:
```diff
@@ -0,0 +1,12 @@
+  versionFunctions: false
+
+functions:
+  writer:
+    handler: core.writer
+  reader:
+    handler: core.reader
```
Let's maybe add comment, that in integration tests VPC and EFS configuration is added dynamically right before test is run.

medikoo reviewed (changes requested):
Thank you @pgrzesik That's outstanding! I've run integration test locally, and it works great.

I've just pointed few last minor issues I've noticed

pgrzesik commented:
Hey @medikoo, thanks a lot for a quick re-review. I addressed the comments + renamed the integration test directory to follow a similar pattern as other ones.

medikoo reviewed (approved):
Thank you @pgrzesik !

We'll publish it tomorrow morning

jimingeorge commented:
Hi @medikoo and @pgrzesik  so would this take care of setting up an efs for us?

pgrzesik commented:
Hello @jimingeorge, sorry for not responding sooner. The answer would be no, it will take care of hooking up existing EFS to your Lambda functions, but you have to create it beforehand. If you're looking for an inspiration on how to do that programatically, you can take a look at the CloudFormation template that is used in integration tests: `tests/integration-all/file-system-config/cloudformation.yml`
